### ISSUE TITLE
Add generic hybrid (KEM) support

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#25884]
In preparation for a possible implementation of the evolving standard https://www.ietf.org/archive/id/draft-kwiatkowski-tls-ecdhe-mlkem-02.html this is to suggest implementing provider-based generic hybrid key management and KEM operation enabling instantiation for the implementation of the standard proposal above.

Design proposal:
- The key manager shall maintain two generic EVP_PKEY structures, incl. public and private key material, e.g., generating two keys instead of one when requested, responding to all key management operations triggering corresponding actions on both objects; the reported security strength should be the maximum of the two constituent security strengths as the KEM logic (see below) must ensure both constituent KEM operations are successful before considering the hybrid operation to be successful; the returned encoded public key shall be the concatenation of the constituent encoded public keys (configurable order).
- The KEM logic shall enable operation of these two EVP_PKEYs as a single KEM in the following way: Do two KEM encap/decap operations instead of one, encapsulating/decapsulating the concatenation of the two shared secrets as per the draft standard above, but be configurable to adapt easily to changes to the draft standard listed above (e.g., order of concatenation).

Code evolution proposal:
Implementation of this may be aided by utilizing the generic code available in https://github.com/openssl/openssl/blob/master/providers/implementations/keymgmt/template_kmgmt.c and https://github.com/openssl/openssl/blob/master/providers/implementations/kem/template_kem.c; any PR implementing this issue is initially invited to be targeted towards the "ml-kem" feature branch in preparation of instantiation with MLKEM768 and classic KEM algorithms as constituent key types. Tests for this generic implementation of course could utilize pre-existing KEM algorithms, too (e.g., P256_X25519), but a focus on X25519MLKEM768 and SecP256r1MLKEM768 is preferable.

Feedback to this proposal solicited, particularly from @paulidale

**Comment by paulidale** [comment]
I'm working on implementing this.  Not sure when it will be complete enough for a draft PR.

Currently it's looking good from the provider algorithm implementation perspective.  The current implementation of a key manager for a combination of X25519 and X448 (for testing purposes only) looks like:

```c
#define NAME_1  X25519
#define NAME_2  X448

#include "hybrid_kmgmt.inc"
```

KEM uses the same approach.

There are, naturally, some complications.

**Comment by baentsch** [comment]
> I'm working on implementing this. 

Thanks very much!

> There are, naturally, some complications.

Yup, ran into some of them myself in case you'd like to discuss. Initial attempts in https://github.com/baentsch/openssl/tree/mb-hybridprovider (if you'd like to see how not to do it :-) Will refrain from proceeding on this for now but would be glad to help your effort as/where you see fit @paulidale

**Comment by tomato42** [comment]
> but a focus on X25519MLKEM768 and SecP256r1MLKEM768 is preferable.

the master branch of the Kwiatkowski draft includes also SecP384r1MLKEM1024

**Comment by tomato42** [comment]
and speaking of the design: [draft-ietf-tls-hybrid-design-11](https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design-11) explicitly allows reuse of the constituent primitive key shares in the ClientHello.

I.e. sending the exact same encapsulation key for ML-KEM-768 when sending key shares for both X25519MLKEM768 and SecP256r1MLKEM768 is explicitly allowed. Same for the X25519 or P-256 key share.

This behaviour is actually already implemented in NSS.

**Comment by t8m** [comment]
> and speaking of the design: [draft-ietf-tls-hybrid-design-11](https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design-11) explicitly allows reuse of the constituent primitive key shares in the ClientHello.
> 
> I.e. sending the exact same encapsulation key for ML-KEM-768 when sending key shares for both X25519MLKEM768 and SecP256r1MLKEM768 is explicitly allowed. Same for the X25519 or P-256 key share.
> 
> This behaviour is actually already implemented in NSS.

I do not think we want to aim for that in the first iteration of the hybrid support. It would be fairly hard to implement on the provider side.

**Comment by baentsch** [comment]
> It would be fairly hard to implement on the provider side.

I honestly wouldn't know how to do this at all staying within the current provider APIs...

Created #25889 to track separately.

**Comment by baentsch** [comment]
> the master branch of the Kwiatkowski draft includes also SecP384r1MLKEM1024

Thanks for the reminder @tomato42 . Indeed, it should be straightforward to instantiate this variant too assuming the provider function sets are suitably "macro"d.

**Comment by tomato42** [comment]
> > and speaking of the design: [draft-ietf-tls-hybrid-design-11](https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design-11) explicitly allows reuse of the constituent primitive key shares in the ClientHello.
> > I.e. sending the exact same encapsulation key for ML-KEM-768 when sending key shares for both X25519MLKEM768 and SecP256r1MLKEM768 is explicitly allowed. Same for the X25519 or P-256 key share.
> > This behaviour is actually already implemented in NSS.
> 
> I do not think we want to aim for that in the first iteration of the hybrid support. It would be fairly hard to implement on the provider side.

I mentioned it now specifically so that we can consider it now and possibly adjust the API

**Comment by t8m** [comment]
> I mentioned it now specifically so that we can consider it now and possibly adjust the API

Yes, thank you for pointing this out. However I do not think there is much to do in terms of API changes on this. Basically it requires special handling in libssl and exporting/importing the individual part keys of the hybrid composite from/to the provider.

I.e., in scenario with 4 key shares (X25519MLKEM768, SecP256r1MLKEM768, X25519, P-256) 1. libssl generates a key for X25519, creates a key share and exports the private key, 2. same for P-256, 3. libssl generates a key for MLKEM768, doesn't create a key share but exports the private key, 4. libssl imports the X25519 and MLKEM768 private key components into the hybrid X25519MLKEM768 EVP_PKEY and creates a key share 5.  libssl imports the P-256 and MLKEM768 private key components into the hybrid SecP256r1MLKEM768 EVP_PKEY and creates a key share. There is not much to be changed in suggested APIs for the generic hybrid KEM support. The import of the private key components can be done with a single EVP_PKEY_fromdata() if the exported OSSL_PARAM arrays are merged by libssl (we have helper functions for that). I do not think we want to add another provider key management import function allowing import from multiple OSSL_PARAM arrays.

As you can see all the special handling is done on the libssl side and there is no need for any extra provider functions. Of course in case the provider does not allow exporting the private key this optimization will not work and such failure needs to be handled gracefully by libssl.

In my opinion this optimization is really low priority and not worth it. We still do not even support sending multiple key shares from the client at all which is a much bigger deficiency.

**Comment by baentsch** [comment]
Adding reference to a related hybrid algorithm spec with a possible impact on the implementation, maybe as a configuration option: https://datatracker.ietf.org/doc/html/draft-kampanakis-curdle-ssh-pq-ke-03 (from https://github.com/openssl/openssl/discussions/26073)

**Comment by petrovr** [comment]
Post-Quantum Traditional (PQ/T) - so far so good but the different standards define specific order .
For instance "PQ/T Hybrid Key Exchange with ML-KEM in SSH" (now in RFC publication queue)  always define PQ first:
- mlkem768nistp256-sha256
- mlkem1024nistp384-sha384
- mlkem768x25519-sha256
 

This is different from TLS where two of algorithms define PQ as second in concatenation. 

----

I cannot understand why "composite" model is if difficult to implement. 

A) use build-in composite scheme
The sample is experimental PKIX-SSH branch HYBRID-DEMO, with four patches(commits) , two groups. First commit in a group is implementation of an algorithm as is provider by cryptographic module and second is correction to match SSH specification. 
The commit "swap keys hack" is self explanatory - it uses  build-in results and swaps the data(bytes) first is PQ(Post-Quantum) and second in T( Traditional) part .

B) use own composition 
This is actual implementation in PKIX-SSH. The related code in source repository is kexhybrid.c and kexkem.c.

----

A) build-in 
This model is limited to build-in functionality i.e., is cannot be used to create a new  composite scheme.

B) own
This model allows program to define own composite scheme based on unitary algorithms from cryptographic modules.
As is visible this is my preference.

----

FIPS

This is the difficult part. The statement is that FIPS mode allows key exchange algorithm PQ || Curve 25519, but disallows  X25519. Also key Ed25519 is allowed.

Generic composite scheme may simplify OpenSSL provider  code.

**Comment by petrovr** [comment]
In brief FIPS regulation blocks use of own composite scheme. 


Now (current main branch) PKIX-SSH code is ready for OpenSSL 3* API, and it is ready for use with FIPS provider even in FIPS mode. 

The key exchange algorithms mlkem768x25519-sha256 is blocked temporary, as X25519 is not usable. I will redesign code to ensure that build-in and based on cryptographic module implementation to be available at the same time and to switch between them depending on running context. This resolves FIPS issue with X25519.

----

Conclusion : It is not difficult  to implement composite scheme using algorithms from different sources.

**Comment by paulidale** [comment]
> In brief FIPS regulation blocks use of own composite scheme.

They don't so long as one algorithm is FIPS validated.  NIST is updating the relevant documents.


> Conclusion : It is not difficult to implement composite scheme using algorithms from different sources.

Feel free to donate an implementation to the project.


The devil is in the details:

- the hybrid gets some get/set params, what do the sub-algorithms get/set from these
- dealing with sub-algorithm param name clashes 
- hybrids with more than two sub-algorithms are anticipated, better support these
- most combination functions should be implemented by OpenSSL, but the user must also be able to supply them
- importing a hybrid key is similar
- converting a hybrid key to/from params which is different to import/export to the key manager
- …
